### DISCUSSION TITLE
elapsed time

### DISCUSSION THREAD

**Discussion by juliermefelix** [serenity-is/Serenity#6134] [Help]
I have a meeting control: parents schedule a meeting, for example, with the Pedagogical Coordinator, the coordinator's secretariat accesses the system and makes the appointment, on the day of the appointment it controls access and informs the entry time and then when it leaves the checkout time and then change the status to closed, but I was asked to make a change to the "Status" or add another column, in order to alert which appointments are in tight times, almost bursting...

The guide would be the time of 01:00h for each Service

* SITUATION 01: When the service is between "00min and 29min", it would have to be highlighted GREEN, to inform that the time is still within the deadline, Showing in the closed place the elapsed time based on the time of entry.
Image example: Entry time 11:40:00 current time: 12:05:00 - So it passed 25 minutes of service so the column of Status would show the time 00:25:00 and in green color.
So successively for the other situations.

The difficulty I'm having is not with the conditional CSS but with the time, I don't know how to change the time.

![image](https://user-images.githubusercontent.com/7516443/145128618-6011eed0-891c-4588-b8a4-f41e4ef3dd76.png)

**Comment by juliermefelix** [comment]
Dear, I'm trying to use moment.js to calculate the hours, but I have problem importing when using import moment from 'moment' visual studio doesn't recognize the 'moment'
![image](https://user-images.githubusercontent.com/7516443/146169763-d2e9f686-6e86-4c8e-9fe9-7f426ccfeb01.png)

**Reply by dfaruque** [reply]
if you just want to calculate hours then the following method will do that.

```
getHours(fromDate: Date, toDate: Date): number {
        let hours = 0;
        if (fromDate && toDate) {
            let totalMiliSeconds = toDate.valueOf() - fromDate.valueOf();
            hours = totalMiliSeconds / (1000 * 60 * 60);
        }
        return hours;
    }
```

**Comment by juliermefelix** [comment]
dfaruque,
Basically that's what I need, but I'm not able to pass the data from the fields of the columns "HoraEntrada" and "HoraSaida" coming from the grid to getHours()
It should basically look like this: getHours(columnGrid1, columnGrid2)
And the result appears in column Grid3

![image](https://user-images.githubusercontent.com/7516443/146372036-7f8f892e-5f98-47ff-8534-d052664b4127.png)

**Comment by juliermefelix** [comment]
Esteemed,
I managed to resolve this issue using Expression SQL. : [Expression("CONVERT(time, DATEADD(s, DATEDIFF(s, t0.[HorarioPedido], t0.[TotalHrs]), CAST(GETDATE() as datetime2)))")]
![image](https://user-images.githubusercontent.com/7516443/146535718-38efb6c0-fe82-4c98-a1d2-d6eb9cc35259.png)

But now I need to update the grid from time to time for example every one minute.
Does anyone have any ideas?
Another thing I would like to put a clock on the button bar.
![image](https://user-images.githubusercontent.com/7516443/146535903-ef344001-5f82-4452-9d5a-1738eb11e626.png)

**Reply by imhemayatsangin** [reply]
Dear! which editor you use for the Time? and in the database what data type do you use? string, datetime which data type?

**Reply by juliermefelix** [reply]
Dear,
I'm not using any editor, and in the database I'm using time(7)

**Reply by imhemayatsangin** [reply]
@juliermefelix Thanks brother, i got it.

**Comment by minhhungit** [comment]
I think you can use slickgrid formatter and call a macro-function for each row (call function inside formatter), something like this

![image](https://user-images.githubusercontent.com/2279508/146653839-58298f88-52df-4f83-9eb0-25446ea9eed7.png)

![image](https://user-images.githubusercontent.com/2279508/146654145-673c9c3d-ea33-4ae3-af98-1ae684f0911f.png)

Demo 
![countdown](https://user-images.githubusercontent.com/2279508/146653918-2a2bb6ba-a880-4b9e-adbf-ba89581f0719.gif)

In above image I use a countdown library and momentjs to make 'countdown' for each order entry, but in your case, I think you instead of using countdown library, you just need to use https://www.w3schools.com/js/js_timing.asp + momentjs , check time in the marco-function. You will need to pass cell ID to the function like above, find element and set result back to cell div.
Hope my sample above will give you some ideas

**Comment by gfo2007** [comment]
@juliermefelix please are you using the [TimeEditor] to decorate your time fields? I keep getting the error below when I use the TimeEditor to decorate my Tim fields.

![Time Editor Error](https://user-images.githubusercontent.com/14317420/199999863-fa5a6172-d247-4c1b-931c-14aa1b582ec9.PNG)

**Reply by juliermefelix** [reply]
Hello,
I didn't use TimeEditor I used custom CSS
![image](https://user-images.githubusercontent.com/7516443/200003874-59ca9d8a-c84c-4fb2-811a-9fb893f87282.png)

![image](https://user-images.githubusercontent.com/7516443/200004073-7d4c082d-20fc-4a6f-9e5b-cf75ccfba4c4.png)

**Reply by gfo2007** [reply]
@juliermefelix thanks for the reply.

What was the data type of your Time fields and on the form, how do your users select the time?

**Reply by juliermefelix** [reply]
in the database I'm using time(7)
 form
![image](https://user-images.githubusercontent.com/7516443/200014259-57aa6754-d2af-4d8f-a925-75689fdbfbdf.png)

**Reply by gfo2007** [reply]
Thanks. Will try this out
